### PR TITLE
Bdd

### PR DESCRIPTION
**Pull request by maninder1438**
multiple filters applied and item reserved.. all tested and working

### CODE REVIEW

File: src/test/java/com/mani/selenium/steps_definations/FilterSteps.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Then("^I should be able to see the products of rating \"([^\"]*)\"$")
+    public void iShouldBeAbleToSeeTheProductsOfRating(double expected) throws Throwable {
+        List<Double> actualRatingList = resultsPage.getAllProductRatings();
+        assertThat(actualRatingList, everyItem(is(greaterThanOrEqualTo(expected))));
+    }
+
+    @And("^Select the price \"([^\"]*)\" from price filter$")
+    public void selectThePriceFromPriceFilter(String price)  {
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/steps_definations/FilterSteps.java:28]: Create separate class for price step def

File: src/test/java/com/mani/selenium/steps_definations/LoginSteps.java [hunk 2]
```diff
@@ -26,15 +20,10 @@ public void i_should_be_able_to_see_account_login_page() throws InterruptedExcep
     }
 
     @And("^the message should be visible \"([^\"]*)\"$")
-    public void theMessageShouldBeVisible(String arg0) throws InterruptedException {
+    public void theMessageShouldBeVisible(String message) throws InterruptedException {
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/steps_definations/LoginSteps.java:22]: Y we need throws ??

File: src/test/java/com/mani/selenium/pages/ResultsPage.java [hunk 3]
```diff
@@ -1,91 +1,114 @@
...
+    private List<WebElement> ratingStar;
+
+    @FindBy(css = ".findability-facet__rating-label")
+    private List<WebElement> ratingWebElements;
+
+    @FindBy(css = ".ProductCardstyles__PriceText-l8f8q8-14.gHrEdF")
+    private List<WebElement> productPrice;
+
+    @FindBy(css = "label[name=\"price\"]")
+    private List<WebElement> priceFilter;
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/pages/ResultsPage.java:27]: Remove \ from locators

File: src/test/java/com/mani/selenium/pages/ResultsPage.java [hunk 4]
```diff
@@ -1,91 +1,114 @@
...
+//                .until(ExpectedConditions.numberOfElementsToBeMoreThan(By.cssSelector("label[name=\"price\"]"), 4));
+              priceFilterShowMore.click();
+            for (WebElement review : priceFilter) {
+                String availableFilter = review.getAttribute("value");
+
+                if (availableFilter.equalsIgnoreCase(selectPrice)) {
+                    review.click();
+                    sleep(6000);
+                    break;
+                }
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/pages/ResultsPage.java]: What should happen when available filter not match with select price ??

File: src/test/java/com/mani/selenium/steps_definations/CommonSteps.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    @When("^I click on account button$")
+    public void i_click_on_account_button(){
+        loginPage.accountLoginPage();
+    }
+    @Given("^I am on account login page$")
+    public void iAmOnAccountLoginPage() {
+        loginPage.accountLoginPage();
+    }
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/steps_definations/CommonSteps.java:33]: Empty line

File: src/test/java/com/mani/selenium/steps_definations/FilterSteps.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    public void iShouldBeAbleToSeeTheProductsHavingPriceBetween(String priceRange)  {
+
+        List<Double> expectedPriceList = resultsPage.getAllProductPrices();
+
+        String[] selectedPrice = priceRange.replace("£","").split("-");
+        double minValuePriceFilter = Double.parseDouble(selectedPrice[0]);
+        double maxValuePriceFilter = Double.parseDouble(selectedPrice[1].trim());
+
+        for (double indPrice : expectedPriceList) {
+            assertThat(indPrice,is(greaterThanOrEqualTo(minValuePriceFilter)));
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/steps_definations/FilterSteps.java:43]: See if 2 assertions can be written in one

Comment (rkesireddy) [reply src/test/java/com/mani/selenium/steps_definations/FilterSteps.java:43]: Hint : both()

File: src/test/java/com/mani/selenium/steps_definations/LoginSteps.java [hunk 7]
```diff
@@ -59,8 +48,8 @@ public void iShouldBeLoggedInSuccessfullyAbleToSeeMyFirstNameOnTopRightCorner(St
         Assert.assertEquals(expectedFirstName,actualFirstName);
     }
     @And("^\"([^\"]*)\" link will be visible on top right corner$")
-    public void linkWillBeVisibleOnTopRightCorner(String arg0)  {
-        Assert.assertEquals(loginPage.signOutButton(),arg0);
+    public void linkWillBeVisibleOnTopRightCorner(String signOutLink)  {
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/steps_definations/LoginSteps.java:50]: Rename method meaningful 

Often link can move position on UI

File: src/test/java/com/mani/selenium/pages/ResultsPage.java [hunk 8]
```diff
@@ -1,91 +1,114 @@
...
+    }
+
+
+    public void selectPriceFilter(String selectPrice) {
+//        new WebDriverWait(driver, 10)
+//                .until(ExpectedConditions.numberOfElementsToBeMoreThan(By.cssSelector("label[name=\"price\"]"), 4));
+              priceFilterShowMore.click();
+            for (WebElement review : priceFilter) {
+                String availableFilter = review.getAttribute("value");
+
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/pages/ResultsPage.java]: Price filter size ==0 what should happen ??

File: src/test/resources/filter.feature [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    Given I am on homepage
+    When I search for a product of "Apple MacBook Air 2019"
+    And Select the product rating "4or more" from the rating filter
+    Then I should be able to see the products of rating "4.0"
+
+  @smoke
+  Scenario: Verify product price filter
+    Given I am on homepage
+    When I search for a product of "phone sim free"
+    And  Select the price "£150 - £200" from price filter
```
Comment (rkesireddy) [inline src/test/resources/filter.feature]: Can you write another scenario combining these 2 scenario and tag it as @e2e

File: src/test/resources/search.feature [hunk 10]
```diff
@@ -7,30 +7,18 @@ Feature: Search
   @smoke
   Scenario: Search for single product
     Given I am on homepage
-    When I search for product "nike"
-    Then I should be able to see "nike" product
+    When I search for a product of "jbjkbkbkjhkbbhbhbjhk5454"
+    Then I should be able to see "nijhbvjhbvhjjhjh5454ke" product
```
Comment (rkesireddy) [inline src/test/resources/search.feature]: Ru expecting any product with that search term??

File: src/test/java/com/mani/selenium/steps_definations/FilterSteps.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import cucumber.api.java.en.Then;
+import cucumber.api.java.en.When;
+
+import java.util.List;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+public class FilterSteps {
+    ResultsPage resultsPage = new ResultsPage();
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/steps_definations/FilterSteps.java:12]: Empty line

File: src/test/java/com/mani/selenium/pages/TrolleyPage.java [hunk 12]
```diff
@@ -1,29 +1,41 @@
...
     public int numberOfProductsinTrolley() {
-        List<WebElement> numberOfProducts = driver.findElements(By.cssSelector(".ProductCard__productLinePrice__3QC7V"));
-        int numberOfProductsInBasket = numberOfProducts.size();
+        int numberOfProductsInBasket = numberOfProductsInTrolley.size();
         return numberOfProductsInBasket;
     }
     public String trolleyTotalProductPrice() {
-        String trolleyTotal = (driver.findElement(By.cssSelector("div[data-e2e=\"basket-total-price\"]")).getText()).replace("£", "").replace(",", "");
+        String trolleyTotal = trollyTotalPrice.getText().replace("£", "").replace(",", "");
         return trolleyTotal;
```
Comment (rkesireddy) [inline src/test/java/com/mani/selenium/pages/TrolleyPage.java:72]: Better !!!! Covert here it self into double n return

**Review by rkesireddy — CHANGES_REQUESTED** [review]
Over good , do chance n let me know

**Comment by maninder1438** [comment]
trying to undo merge

**Pull request merged**